[For discussion] Run integration test with multiple servers

daviddrysdale opened the pull request:
Multiple CT personalities, multiple Log RPC servers, with random "load balancing" between them.

codingllama commented on server/toyloglb/main.go:
```diff
@@ -0,0 +1,201 @@
+	bc := lb.pick()
+	glog.V(3).Infof("forward QueueLeaves request to backend %s", bc.server)
+	return bc.client.QueueLeaves(ctx, req)
+}
```
nit: Add a blank line between methods?

daviddrysdale replied:
Done.

daviddrysdale commented on integration/ct_config.sh:
```diff
@@ -0,0 +1,21 @@
+# Builds configuration for a CT test, set in ${CT_CFG},
+# sets ${TREE_IDS} and provisions logs for them.
+set -e
+INTEGRATION_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Done, across all the scripts.

daviddrysdale commented on integration/ct_config.sh:
```diff
@@ -0,0 +1,21 @@
+. ${INTEGRATION_DIR}/common.sh
+
+# Build config file with absolute paths
+CT_CFG=$(mktemp ${INTEGRATION_DIR}/ct-XXXXXX)
```
Done, across all the scripts.

daviddrysdale commented on integration/ct_config.sh:
```diff
@@ -0,0 +1,21 @@
+# Build config file with absolute paths
+CT_CFG=$(mktemp ${INTEGRATION_DIR}/ct-XXXXXX)
+sed "s!@TESTDATA@!${TESTDATA}!" ./integration/ct_integration_test.cfg > ${CT_CFG}
+trap "rm ${CT_CFG}" EXIT
```
Done.

daviddrysdale commented on integration/ct_config.sh:
```diff
@@ -0,0 +1,21 @@
+trap "rm ${CT_CFG}" EXIT
+
+# Retrieve tree IDs from config file
+TREE_IDS=$(grep LogID ${CT_CFG} | grep -o '[0-9]\+'| xargs)
```
Fixed.

daviddrysdale commented on integration/ct_config.sh:
```diff
@@ -0,0 +1,21 @@
+for id in ${TREE_IDS}
+do
+    echo "Provisioning test log (Tree ID: ${id}) in database"
+    ${SCRIPTS_DIR}/wipelog.sh ${id}
```
Done.

daviddrysdale commented on integration/ct_integration_test.go:
```diff
@@ -30,7 +31,7 @@ import (
 )
 
-var httpServerFlag = flag.String("ct_http_server", "localhost:8092", "Server address:port")
+var httpServerFlag = flag.String("ct_http_server", "localhost:8092", "Comma-separated list of servers, each as address:port")
```
Done.

daviddrysdale replied:
Done.

daviddrysdale commented on integration/ct_integration_test.go:
```diff
@@ -517,35 +531,51 @@ func (want *wantStats) done(ep string, rc int) {
+		}
+		var stats ctfe.AllStats
+		if err := json.NewDecoder(httpRsp.Body).Decode(&stats); err != nil {
+			return fmt.Errorf("failed to json.Decode() result: %v", err)
```
Done.

daviddrysdale commented on integration/ct_integration_test.go:
```diff
@@ -517,35 +531,51 @@ func (want *wantStats) done(ep string, rc int) {
+		// Accumulate per-server stats for this log into overall per-log stats.
+		got.HTTPAllReqs += gotSingle.HTTPAllReqs
+		for _, rc := range rcs {
+			got.HTTPAllRsps[rc] += gotSingle.HTTPAllRsps[rc]
```
These are different statuses -- line 551 detects whether we fail to retrieve stats, whereas the range for all of the results that the server has served (e.g. it will have a 400 for the add-chain of a corrupt cert).

daviddrysdale commented on integration/ct_integration_test.sh:
```diff
@@ -6,19 +6,7 @@ INTEGRATION_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
-    ${SCRIPTS_DIR}/wipelog.sh ${id}
-    ${SCRIPTS_DIR}/createlog.sh ${id}
-done
+. ${INTEGRATION_DIR}/ct_config.sh
```
Done.

daviddrysdale commented on integration/ct_multi_test.sh:
```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
+set -e
+INTEGRATION_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Done across `integration/*sh`

daviddrysdale commented on integration/ct_multi_test.sh:
```diff
@@ -0,0 +1,87 @@
+. ${INTEGRATION_DIR}/ct_config.sh
+
+# Start a set of Log RPC servers.  Note that each of them will run their own
+# sequencer; a proper deployment should have a single master sequencer, but
```
Added a TODO to update the comment.

daviddrysdale commented on integration/ct_multi_test.sh:
```diff
@@ -0,0 +1,87 @@
+./toyloglb --backends ${RPC_SERVERS} --port ${LB_PORT} &
+LB_SERVER_PID=$!
+popd > /dev/null
+trap "kill -INT ${LB_SERVER_PID} ${RPC_SERVER_PIDS}" EXIT
```
Doesn't `trap` entirely replace the action for the given signal?

codingllama replied:
You're right, my bad.

daviddrysdale commented on integration/ct_multi_test.sh:
```diff
@@ -0,0 +1,87 @@
+set -e
+
+rm ${CT_CFG}
+trap - EXIT
```
Done

daviddrysdale commented on server/trillian_log_server/main.go:
```diff
@@ -125,18 +125,17 @@ func main() {
+		glog.Infof("Creating HTP server starting on port: %d", *httpPortFlag)
 		err := startHTTPServer(*httpPortFlag)
-
 		if err != nil {
```
Done.

daviddrysdale commented on integration/integration_test.sh:
```diff
@@ -5,3 +5,4 @@ INTEGRATION_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 runTest map_integration_test.sh "Map integration test"
 runTest log_integration_test.sh "Log integration test"
 runTest ct_integration_test.sh "CT integration test"
+runTest ct_multi_test.sh "CT multi-server integration test"
```
I could go either way -- you're right that it's a superset, but also I can imagine a situation where the 1+1 test works but the 3+3 test fails, and that would be helpful information.

How about having a single script, but with parameters to indicate how many of each to run, and then have the overall script run:
```
runTest "ct_integration_test.sh 1 1" "CT integration test"
runTest "ct_integration_test.sh 3 3" "CT multi-server integration test"
```
?

(If we go for this, I'll do a separate pull request for it)

codingllama replied:
SGTM.

daviddrysdale commented:
"Should we simply have a CT personality load balancer instead of adding the complexity to the test?"

I don't think so.  A real deployment will normally have some sort of LB/reverse proxy, so the LB would only be needed for the test; more significantly it would have to do the multi-backend statistics summing that's in the integration test.  So I lean towards just keeping the LB code in the integration test itself.

codingllama commented:
RE: "Should we simply have a CT personality load balancer instead of adding the complexity to the test?"

I meant it as a toy load balancer, just for tests, like we have now between Trillian and personalities. Given that we already have one, we might as well add a 2nd "toylb" and simplify ct_multi_test / ct_integration_test.go (no need to pick random clients or have multiple urls, for example). To keep tests simple, we could have the LB automatically sum stats from all frontends.

daviddrysdale commented:
PTAL -- I've:
 - rebased everything up to your approval to get a better commit sequence
 - changed my mind about the location of `server/toyloglb` and `git mv`ed it to `testonly/loglb` (and folded that in with a rebase to make myself look less indecisive)
 - added a commit to sort out `trap` processing
 - added two commits to commonize `ct_integration_test.sh` & `ct_multi_test.sh`.

I haven't pulled out the HTTP load balancing stuff from `ct_integration_test.go`; I'm still in two minds about that change as the stats summation gets rather fiddlier in the general case (as there's no specific log ID to sum for).

codingllama commented on integration/ct_integration_test.sh:
```diff
@@ -1,62 +1,113 @@
+LB_PORT=46962
 
-. ${INTEGRATION_DIR}/ct_config.sh
+for i in $(seq 1 ${RPC_SERVER_COUNT})
```
Suggestion: Simplify to:

for ((i=0; i < RPC_SERVER_COUNT; i++))); do ...; done

daviddrysdale replied:
Done.

codingllama commented on integration/ct_integration_test.sh:
```diff
@@ -1,62 +1,113 @@
-. ${INTEGRATION_DIR}/ct_config.sh
+for i in $(seq 1 ${RPC_SERVER_COUNT})
+do
+  port=$(expr ${BASE_RPC_PORT} + ${i})
```
Suggestion: port=$((BASE_RPC_PORT + i))

daviddrysdale replied:
Done.

codingllama commented on integration/ct_integration_test.sh:
```diff
@@ -1,62 +1,113 @@
+for i in $(seq 1 ${RPC_SERVER_COUNT})
+do
+  port=$(expr ${BASE_RPC_PORT} + ${i})
+  if [ ${i} == 1 ]; then
```
nit: Avoid "["?

if [[ $i -eq 1 ]]; then ...; fi

daviddrysdale replied:
Done.

codingllama commented on integration/ct_integration_test.sh:
```diff
@@ -1,62 +1,113 @@
 
-echo "Starting Log RPC server on port ${RPC_PORT}"
-pushd ${TRILLIAN_ROOT} > /dev/null
+for i in $(seq 1 ${HTTP_SERVER_COUNT})
```
Same suggestions as the block above.

daviddrysdale replied:
Done.

daviddrysdale commented on integration/common.sh:
```diff
@@ -1,19 +1,19 @@
-  echo "=== RUN   ${2}"
-  ${INTEGRATION_DIR}/$1
+  echo "=== RUN   ${1}"
+  "${INTEGRATION_DIR}"/"$2" "$3" "$4" "$5" "$6" "$7" "$8"
```
Done for 1/2, but 3 onwards are going to be test-specific so I can't think of a sensible way to name them.

daviddrysdale commented on integration/common.sh:
```diff
@@ -26,3 +26,18 @@ function waitForServerStartup() {
+
+# Clean up anything in ${TO_KILL} and ${TO_DELETE}.
+function onExit() {
+  if [ "${TO_KILL}" != "" ]; then
```
Done.

daviddrysdale commented on integration/common.sh:
```diff
@@ -26,3 +26,18 @@ function waitForServerStartup() {
+  fi
+  if [ "${TO_DELETE}" != "" ]; then
+    echo "Deleting ${TO_DELETE} on exit"
+    rm ${TO_DELETE}
```
Won't that fail if `TO_DELETE` has multiple files in it?

daviddrysdale commented on integration/ct_integration_test.sh:
```diff
@@ -1,62 +1,113 @@
-RPC_PORT=36962
-CT_PORT=6962
+# Default to one RPC server and one HTTP server.
+RPC_SERVER_COUNT=${1:-1}
```
Not sure I follow -- isn't that what I'm doing?

codingllama replied:
Yes, sorry for the noise.

daviddrysdale commented:
Thanks for the bash education (I'm used to writing things for plain old `/bin/sh`)!

codingllama commented:
You're welcome! I've picked up some new things here too! :)
